designs/cloudstack-multiple-endpoints.md

```diff
@@ -0,0 +1,177 @@
+//
+Domain string `json:"domain"`
+// Zones is a list of one or more zones that are managed by a single CloudStack management endpoint.
+Zone CloudStackZone `json:"zone"`
+// Account typically represents a customer of the service provider or a department in a large organization. Multiple users can exist in an account, and all CloudStack resources belong to an account. Accounts have users and users have credentials to operate on resources within that account. If an account name is provided, a domain must also be provided.
+Account string `json:"account,omitempty"`
+// CloudStack Management API endpoint's IP. It is added to VM's noproxy list
+ManagementApiEndpoint string `json:"managementApiEndpoint"`
+```
+
+and we would parse these resources and pass them into CAPC by modifying the templates we have currently implemented. We can then use this new model to read in credentials, perform pre-flight checks, plumb data to CAPC, and support upgrades in the controller. The goal would be to make these new resources backwards compatible via code
+
+### Failure Domain
+
+A failure domain is a CAPI concept which serves to improve HA and availability by destributing machines across "failure domains", as discussed [here](https://cluster-api.sigs.k8s.io/developer/providers/v1alpha2-to-v1alpha3.html?highlight=domain#optional-support-failure-domains). 
+CAPC currently utilizes them to distribute machines across CloudStack Zones. However, we now want to go a step further to support our customer and consider the following unique combination to be a failure domain:
+
+1. Cloudstack endpoint
+2. Cloudstack domain
+3. Cloudstack zone
+4. Cloudstack account
+
+You can find more information about these Cloudstack resources [here](http://docs.cloudstack.apache.org/en/latest/conceptsandterminology/concepts.html#cloudstack-terminology)
+
+### `CloudstackDatacenterConfig` Validation
+
+With the multi-endpoint system for the Cloudstack provider, customers reference a CloudstackMachineConfig and it's created across multiple failure domains. The implication
+is that all the Cloudstack resources such as image, ComputeOffering, ISOAttachment, etc. must be available in *all* the failure domains, or all the Cloudstack endpoints,
+and these resources must be referenced by name, not unique ID. This would mean that for each CloudstackMachineConfig, we have to make sure that all the prerequisite
+Cloudstack resources are available in all the Cloudstack API endpoints. 
+
+In practice, the pseudocode would look like:
+
+for failureDomain in failureDomains:
+  for machineConfig in machineConfigs:
+    validate resource presence with the failureDomain's instance of the CloudMonkey executable
+
+### Cloudstack credentials
+
+
+In a multi-endpoint Cloudstack cluster, each endpoint may have its own credentials. We propose that Cloudstack credentials will be passed in via environment variable in the same way as they are currently,
```

Comment:
what abt user specifies a file path to the .ini and pass it as ENV?

Reply:
I think we could do that as well if you prefer. Although it might require some changes in the e2e test framework. I think there we're using the encoded value instead of any specific file. I'm not sure how/if we could download the file locally and then use it in the e2e tests but it does seem easier for the customer.

@vivek-koppuru any thoughts on this? I know we've discussed how to have customers pass data in at length.

Reply:
for snow we pass in file path as ENV vars

Reply:
From some discussions internally, it feels like storing credentials in a transient environment variable is more secure than in a file written to disk. That is why we prefer to go the env var route, although it is slightly more effort for the customer